cmd/tetra/debug/maps.go

```diff
@@ -0,0 +1,235 @@
+should be empty.
+
+The fourth table aggregate map by name and add the count column indicating how
+many maps were aggregated for the given name. It ranks map from the one
+consuming the most memory to the one consuming the less. Use the -n flag to
+adjust the number of item in the table.
+
+[^1]: https://lore.kernel.org/all/20230305124615.12358-1-laoar.shao@gmail.com/`, tetragonBPFFS),
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			// check that the bpffs exists and we have permissions
+			_, err := os.Stat(tetragonBPFFS)
+			if err != nil {
+				return fmt.Errorf("make sure tetragon is running and you have enough permissions: %w", err)
+			}
+
+			// print BPF maps memory usage
+			allMapsMem, err := bugtool.TotalByteMemlockAllMaps()
+			if err != nil {
+				return fmt.Errorf("failed to retrieve all maps memlock: %w", err)
+			}
+			pinnedProgsMapsMem, err := bugtool.TotalByteMemlockMapsFromProgs(tetragonBPFFS)
+			if err != nil {
+				return fmt.Errorf("failed to retrieve maps from pinned prog memlock: %w", err)
+			}
+			pinnedMapsMem, err := bugtool.TotalByteMemlockMaps(tetragonBPFFS)
+			if err != nil {
+				return fmt.Errorf("failed to retrieve pinned maps memlock: %w", err)
+			}
+			w := tabwriter.NewWriter(cmd.OutOrStdout(), 0, 0, 3, ' ', 0)
+			fmt.Fprintln(w, "AllMaps\tPinnedProgsMaps\tPinnedMaps")
+			fmt.Fprintf(w, "%d\t%d\t%d\n", allMapsMem, pinnedProgsMapsMem, pinnedMapsMem)
+			w.Flush()
+			cmd.Println()
+
+			// print details on map distribution
+			pinnedProgsMapsSet := map[int]bugtool.ExtendedMapInfo{}
+			pinedProgsMaps, err := bugtool.FindMapsUsedByPinnedProgs(tetragonBPFFS)
+			if err != nil {
+				return fmt.Errorf("failed to retrieve maps used by pinned progs: %w", err)
+			}
+			for _, info := range pinedProgsMaps {
+				id, ok := info.ID()
+				if !ok {
+					return errors.New("failed retrieving progs ID, need >= 4.13, kernel is too old")
+				}
+				pinnedProgsMapsSet[int(id)] = info
+			}
+
+			pinnedMapsSet := map[int]bugtool.ExtendedMapInfo{}
+			pinnedMaps, err := bugtool.FindPinnedMaps(tetragonBPFFS)
+			if err != nil {
+				return fmt.Errorf("failed to retrieve pinned maps: %w", err)
+			}
+			for _, info := range pinnedMaps {
+				id, ok := info.ID()
+				if !ok {
+					return errors.New("failed retrieving map ID, need >= 4.13, kernel is too old")
+				}
+				pinnedMapsSet[int(id)] = info
+			}
+
+			w = tabwriter.NewWriter(cmd.OutOrStdout(), 0, 0, 3, ' ', 0)
+			fmt.Fprintln(w, "PinnedProgsMaps\tPinnedMaps\tInter\tExter\tUnion\tDiff")
+			fmt.Fprintf(w, "%d\t%d\t%d\t%d\t%d\t%d\n",
+				len(pinnedProgsMapsSet),
+				len(pinnedMapsSet),
+				len(inter(pinnedMapsSet, pinnedProgsMapsSet)),
+				len(exter(pinnedMapsSet, pinnedProgsMapsSet)),
+				len(union(pinnedMapsSet, pinnedProgsMapsSet)),
+				len(diff(pinnedMapsSet, pinnedProgsMapsSet)),
+			)
+			w.Flush()
+			cmd.Println()
+
+			// print details on the diff
+			w = tabwriter.NewWriter(cmd.OutOrStdout(), 0, 0, 3, ' ', 0)
```

Comment:
I'm getting empty diff with just column output which is confusing,
```
PinnedProgsMaps   PinnedMaps   Inter   Exter   Union   Diff
62                31           31      31      62      0

ID   Name   Type   KeySize   ValueSize   MaxEntries   Memlock

Name              Type          KeySize   ValueSize   MaxEntries   Count   TotalMemlock   PercentOfTotal
execve_map        Hash          4         632         32768        1       23072416       83.6%
```

could we just say there's no diff or don't print those headers at all?

Reply:
agree but I then wonder how to describe the content of this output since I rely on first, second, third, etc table to describe it. Maybe printing there's no diff then in place of the table 🤔

Reply:
it will now print "Empty diff table" and "Empty BPF memory consumption table" in place of the two last tables if they are empty.